flink-core/src/test/java/org/apache/flink/testutils/ThrowOnDoubleMemoryFreeing.java

```diff
@@ -19,32 +19,28 @@
 package org.apache.flink.testutils;
 
 import org.apache.flink.core.memory.MemorySegment;
-import org.apache.flink.core.testutils.CommonTestUtils;
 import org.apache.flink.util.ExternalResource;
 
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.Properties;
 
 /**
  * A rule that can be used to enable throwing exception on multiple freeing of {@link
  * MemorySegment}.
  */
 public class ThrowOnDoubleMemoryFreeing implements ExternalResource {
 
-    private Map<String, String> systemEnv = new HashMap<>();
+    private Properties systemProperties;
 
     @Override
     public void before() throws Exception {
-        systemEnv = System.getenv();
-        CommonTestUtils.setEnv(
-                Collections.singletonMap(MemorySegment.CHECK_MULTIPLE_FREE_PROPERTY, ""), false);
+        systemProperties = System.getProperties();
+        System.setProperty(MemorySegment.CHECK_MULTIPLE_FREE_PROPERTY, "");
```

Comment:
does this approach actually work when running the test through maven?
Once the MemorySegment was loaded once without this property being set, modifications to the property will have no effect.

it would be good to have a test that ensures this actually works as expected.

Reply:
Currently we have `MemoryManagerTest` using this.
I think this should work fine as a class rule, but may not work as a test rule.
This is intermediate anyway. Once we have fixed existing multiple-free cases, we will set the property for all the CI tests, and will not need this rule anymore.

Reply:
But the `MemorySegment` class could already have been loaded by another test class?

I don't think unit tests have isolated class-loaders.

Reply:
Double-checking the maven surefire plugin documents, I think you are right. No isolation is guaranteed between unit tests.

This approach came from this [PR discussion](https://github.com/apache/flink/pull/15246#discussion_r595776849). We are not ready to fail on double memory freeing for all the tests yet. And if it doesn't fail, the test cases added in the PR does not protect anything.

Maybe we can remove `ThrowOnDoubleMemoryFreeing`, and leave the test cases to only take effect when the failings are activated. We may use `Assume` to ignore the cases when the property if not set.

cc @dawidwys

Reply:
That's a very good point! The current solution might break other tests :(

I would not add the `Assume` as we do not need to ignore the tests. The tests will still pass even without the throwing the exception. The problem is that without the exception they would pass even without the fix in place, but I guess we can live with it for now.